pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go

```diff
@@ -44,44 +45,130 @@ var CPUUsageAllowance = settings.RegisterDurationSetting(
+		provider:        provider,
+		responseChan:    make(chan *roachpb.TokenBucketResponse, 1),
+		lowRUNotifyChan: make(chan struct{}, 1),
+	}
+	c.limiter.Init(c.timeSource, c.lowRUNotifyChan)
+
+	// TODO(radu): support changing the tenant configuration at runtime.
+	c.costCfg = tenantcostmodel.ConfigFromSettings(&st.SV)
+	return c, nil
+}
+
 // NewTenantSideCostController creates an object which implements the
 // server.TenantSideCostController interface.
 func NewTenantSideCostController(
 	st *cluster.Settings, tenantID roachpb.TenantID, provider kvtenant.TokenBucketProvider,
 ) (multitenant.TenantSideCostController, error) {
-	if tenantID == roachpb.SystemTenantID {
-		return nil, errors.AssertionFailedf("cost controller can't be used for system tenant")
-	}
-	tc := &tenantSideCostController{
-		settings: st,
-		tenantID: tenantID,
-		provider: provider,
-	}
-	tc.mu.costCfg = tenantcostmodel.ConfigFromSettings(&st.SV)
-	sv := &st.SV
-	tenantcostmodel.SetOnChange(sv, func(context.Context) {
-		tc.mu.Lock()
-		defer tc.mu.Unlock()
-		tc.mu.costCfg = tenantcostmodel.ConfigFromSettings(sv)
-	})
-	return tc, nil
+	return newTenantSideCostController(st, tenantID, provider, timeutil.DefaultTimeSource{})
+}
+
+// TestingTenantSideCostController is a testing variant of
+// NewTenantSideCostController which allows using a specified TimeSource.
+func TestingTenantSideCostController(
+	st *cluster.Settings,
+	tenantID roachpb.TenantID,
+	provider kvtenant.TokenBucketProvider,
+	timeSource timeutil.TimeSource,
+) (multitenant.TenantSideCostController, error) {
+	return newTenantSideCostController(st, tenantID, provider, timeSource)
 }
 
 func init() {
 	server.NewTenantSideCostController = NewTenantSideCostController
 }
 
 type tenantSideCostController struct {
-	settings *cluster.Settings
-	tenantID roachpb.TenantID
-	provider kvtenant.TokenBucketProvider
+	timeSource timeutil.TimeSource
+	settings   *cluster.Settings
+	costCfg    tenantcostmodel.Config
+	tenantID   roachpb.TenantID
+	provider   kvtenant.TokenBucketProvider
+	limiter    limiter
+	stopper    *stop.Stopper
+	cpuSecsFn  multitenant.CPUSecsFn
 
 	mu struct {
 		syncutil.Mutex
 
-		costCfg     tenantcostmodel.Config
 		consumption roachpb.TenantConsumption
 	}
+
+	// lowRUNotifyChan is used when the number of available RUs is running low and
+	// we need to send an early token bucket request.
+	lowRUNotifyChan chan struct{}
+
+	// responseChan is used to receive results from token bucket requests, which
+	// are run in a separate goroutine. A nil response indicates an error.
+	responseChan chan *roachpb.TokenBucketResponse
+
+	// run contains the state that is updated by the main loop.
+	run struct {
+		now         time.Time
+		cpuSecs     float64
+		consumption roachpb.TenantConsumption
+
+		// TargetPeriodSetting value at the last update.
+		targetPeriod            time.Duration
+		requestInProgress       bool
+		initialRequestCompleted bool
+		requestNeedsRetry       bool
```

Comment:
some commentary about how `requestNeedsRetry` interacts with `requestInProgress` might be handy. Very nit-picky but I'd put initialRequestCompleted above `requestInProgress` because those other two interact with each othe.